packages/core/test/transient.test.ts

```diff
@@ -398,4 +398,42 @@ describe('transient states (eventless transitions)', () => {
     const state = machine.transition('a', 'FOO');
     expect(state.value).toBe('pass');
   });
+
+  it('should work with transient transition on root', done => {
+    const machine = createMachine<any, any>({
+      id: 'machine',
+      initial: 'first',
+      context: { count: 0 },
+      states: {
+        first: {
+          on: {
+            ADD: {
+              actions: assign({ count: ctx => ctx.count + 1 })
+            }
+          }
+        },
+        success: {
+          type: 'final'
+        }
+      },
+      on: {
+        '': [
+          {
+            target: '.success',
```

Comment:
Without adding the check to see if the machine was "done", this could potentially infinite loop.

The developer needs to make sure proper checks in `cond` are in place to prevent infinite loops when using transient transitions.

Reply:
We could add an explicit test to guard us against introducing that infinite loop in the future. It could be a copy of this machine here - we'd have to move moving from "first" state to "success" state to the inner level (from the root) and have a "dangerous" transient state on the root.

Reply:
Yep I plan to add heuristics for this in a follow-up PR